package/MDAnalysis/analysis/base.py

```diff
@@ -430,25 +471,231 @@ def run(self, start=None, stop=None, step=None, frames=None,
         verbose = getattr(self, '_verbose',
                           False) if verbose is None else verbose
 
-        self._setup_frames(self._trajectory, start=start, stop=stop,
-                           step=step, frames=frames)
         logger.info("Starting preparation")
-        self._prepare()
         logger.info("Starting analysis loop over %d trajectory frames",
                     self.n_frames)
 
-        for i, ts in enumerate(ProgressBar(
-                self._sliced_trajectory,
+        frame_indices, trajectory = self._bslices[bslice_idx]
+        for idx, ts in enumerate(ProgressBar(trajectory,
                 verbose=verbose,
                 **progressbar_kwargs)):
+            i = frame_indices[idx]
             self._frame_index = i
             self._ts = ts
             self.frames[i] = ts.frame
             self.times[i] = ts.time
             self._single_frame()
         logger.info("Finishing up")
+        return self
+
+    def _setup_bslices(self, start=None, stop=None, step=None, frames=None): 
+        """
+        Set the self._bslices (the workload distribution scheme) for the future delayed
+        computations.
+
+        Parameters
+        ----------
+        start : int, optional
+            start frame of analysis
+        stop : int, optional
+            stop frame of analysis
+        step : int, optional
+            number of frames to skip between each analysed frame
+        frames : array_like, optional
+            array of integers or booleans to slice trajectory; `frames` can
+            only be used *instead* of `start`, `stop`, and `step`. Setting
+            *both* `frames` and at least one of `start`, `stop`, `step` to a
+            non-default value will raise a :exc:`ValueError`.
+
+        Raises
+        ------
+        ValueError
+            if *both* `frames` and at least one of `start`, `stop`, or `frames`
+            is provided (i.e., set to another value than ``None``)
+
+        Returns
+        -------
+        bslices : list of (bstart, bstop, bstep, bframes) tuples.
+            Iterator will have size of self._n_bslices.
+        """
+        if frames is not None:
+            if not all(opt is None for opt in [start, stop, step]):
+                raise ValueError("start/stop/step cannot be combined with "
+                                 "frames")
+
+        else:  # frames is None
+            start, stop, step = self._trajectory.check_slice_indices(start, stop, step)
+            frames = list(range(start, stop, step))
+        
+        self.n_frames_total = len(frames)
+        n_bslices = self._n_bslices
+
+        def array_is_boolean(arr):
```

Comment:
Defining a function here increase code complexity IMO. Is it possible to move it elsewhere? A possibility is a private function of the module, but since the name is very clear and it is quite general, moving it to `MDAnalysis.lib.util` might be a good alternative too?

Reply:
If you move this, add a docstring. And maybe the name "array" is too limiting since the coercion to a set would work with many different collections. `collection_is_boolean` or similar might be better?

Reply:
I re-wrote it as a comprehension -- I believe there's no need in a function now.